File: java/src/com/db4o/dg2db4o/chapter12/MixedExample.java
/*
 * Mixed.java
 *
 * Created on 02 April 2006, 20:29
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

package com.db4o.dg2db4o.chapter12;

import com.db4o.*;
import com.db4o.ext.*;
import java.util.Scanner;

/**
 *
 * @author jpa1
 */
public class MixedExample {
    private final String DBFILENAME = "c:/mixed.yap";
    
    public static void main(String[] args){
        MixedExample mixed = new MixedExample();
        mixed.convertClassNames();
        mixed.readAll();
        mixed.addData();
        mixed.readAll();
    }
    
    public void readAll() {
        System.out.println("Reading database in Java");
        ObjectContainer db = Db4o.openFile(DBFILENAME);
        ObjectSet results = db.get(new Person());
        while (results.hasNext()) {
            System.out.println(results.next());
        }
        db.close();
    }
    
    public void addData() {
        boolean stop = false;
        System.out.println("Adding new data in Java");
        ObjectContainer db = Db4o.openFile(DBFILENAME);
        Scanner sc = new Scanner(System.in);
        while (!stop) {
            System.out.print("\nEnter the developer's name and age (e.g. 'Tom 44'): ");
            String name = sc.next();
            if(name.equals("stop"))
                stop = true;
            else {
                db.set(new Person(name, sc.nextInt()));
                db.commit();
            };
        }
        db.close();
    }
    
    public void convertClassNames() {
        System.out.println("Checking class names in Java");
        ObjectContainer db = Db4o.openFile(DBFILENAME);
        StoredClass[] classes = db.ext().storedClasses();
        for (int i = 0; i < classes.length; i++) {
            StoredClass storedClass = classes[i];
            String name = storedClass.getName();
            //System.out.println(name);
            String newName = null;
            int pos = name.indexOf(",");
            if(pos > 0){
                newName = name.substring(0, pos);
                storedClass.rename(newName);
                System.out.println("Renaming " + name + " to " + newName);
            }
        }
        db.close();
    }
    
}

File: csharp/MixedExample/MixedExample.cs
using System;
using com.db4o;
using com.db4o.ext;

namespace com.db4o.dg2db4o.chapter12
{
    class MixedExample
    {
        private string DBFILENAME = "c:/mixed.yap";

        public static void Main(String[] args)
        {
            MixedExample mixed = new MixedExample();
            mixed.ConvertClassNames("com.db4o.dg2db4o.chapter12", "MixedExample");  // use assembly name created in this project, was dg2db4o in book
            mixed.ConvertClassNames("com.db4o", "db4o");
            mixed.ReadAll();
            mixed.AddData();
            mixed.ReadAll();

            Console.WriteLine("--------------");
            Console.WriteLine("Press ENTER to end");
            Console.ReadLine();

        }

        public void ReadAll()
        {
            Console.WriteLine("Reading database in .NET");
            ObjectContainer db = Db4o.OpenFile(DBFILENAME);
            ObjectSet results = db.Get(new Person());
            while (results.HasNext())
            {
                Console.WriteLine(results.Next());
            }
            db.Close();
        }

        public void AddData()
        {
            bool stop = false;
            Console.WriteLine("Adding new data in .NET");
            ObjectContainer db = Db4o.OpenFile(DBFILENAME);

            while (!stop)
            { // Your stop condition here...
                Console.WriteLine("Please enter a name: ");
                String name = Console.ReadLine();
                if (name.Equals("stop"))
                    break;
                Console.WriteLine("Please enter an age: ");
                int age = Convert.ToInt32(Console.ReadLine());
                Person p = new Person(name, age);
                db.Set(p);
            }
            db.Close();
        }

        public void ConvertClassNames(string classNamespace, string assembly)
        {
            Console.WriteLine("Checking class names in .NET");
            ObjectContainer db = Db4o.OpenFile(DBFILENAME);
            StoredClass[] classes = db.Ext().StoredClasses();
            for (int i = 0; i < classes.Length; i++)
            {
                StoredClass storedClass = classes[i];
                String name = storedClass.GetName();
                //Console.WriteLine(name);
                String newName = null;
                int pos = name.IndexOf(",");
                if (pos == -1)
                {
                    pos = name.IndexOf(classNamespace);
                    if (pos == 0)
                    {
                        newName = name + ", " + assembly;
                        storedClass.Rename(newName);
                        Console.WriteLine("Renaming " + name + " to " + newName);
                    }
                }
            }
            db.Close();
        }
    }
}
